### PR TITLE
Feature/em 49 update code climate information

### PR DESCRIPTION
**Pull request by mrubi-rootstrap**
## What does this PR do?

It implements the update of the CodeClimate information for each project [EM-49](https://rootstrap.atlassian.net/jira/software/projects/EM/boards/42?selectedIssue=EM-49)

- It creates the rake task `code_climate:update` that updates CodeClimate information for all existing projects
- It creates the class UpdateAllProjectsService that updates CodeClimate information for all existing projects
- It creates the class UpdateProjectService that updates CodeClimate information for a single project
- It creates the class GetProjectInfo that fetches and gathers the CodeClimate information for a single project into a ProjectSummary class
- It wraps the CodeClimate API with the classes in `app/services/code_climate/api`

### CODE REVIEW

File: app/services/code_climate/api/object.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+module CodeClimate
+  module Api
+    class Object
```
Comment (hdamico) [inline app/services/code_climate/api/object.rb:3]: I would try to find another name for this class, Ruby already has a class called Object, seems odd for me

File: app/services/code_climate/api/remote_query.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+module CodeClimate
+  module Api
+    class RemoteQuery
```
Comment (hdamico) [inline app/services/code_climate/api/remote_query.rb:3]: What is this class for?

File: app/services/code_climate/api/repository.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+module CodeClimate
+  module Api
+    class Repository < CodeClimate::Api::Object
+      def summary
+        default_branch_most_recent_snapshot&.summary
+      end
+
+      def id
```
Comment (hdamico) [inline app/services/code_climate/api/repository.rb:8]: I would change the name of the method with something like, `id_retriever` or `repository_id`

File: app/services/code_climate/api/object.rb (hunk 1 above)
Comment (mrubi-rootstrap) [reply app/services/code_climate/api/object.rb:3]: Yes, I didn't know what to use. `CodeClimate::Api::Resource` could be better?

File: spec/services/code_climate/code_climate_spec_helper.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
+require 'rails_helper'
+
+module CodeClimateSpecHelper
+  def base_url
```
Comment (hdamico) [inline spec/services/code_climate/code_climate_spec_helper.rb:4]: Wouldn't a constant be better?

File: app/services/code_climate/api/remote_query.rb (hunk 2 above)
Comment (mrubi-rootstrap) [reply app/services/code_climate/api/remote_query.rb:3]: It encapsulates the `url` and the parameters of an endpoint such as `.../repos?github_slug='rootstrap/rs-code-review-metrics'

File: app/services/code_climate/api/repository.rb (hunk 3 above)
Comment (mrubi-rootstrap) [reply app/services/code_climate/api/repository.rb:8]: Will change it. Thanks!

File: spec/services/code_climate/code_climate_spec_helper.rb (hunk 4 above)
Comment (mrubi-rootstrap) [reply spec/services/code_climate/code_climate_spec_helper.rb:4]: Indeed. Will fix it. Thanks!

File: app/services/code_climate/api/object.rb (hunk 1 above)
Comment (mrubi-rootstrap) [reply app/services/code_climate/api/object.rb]: Changed it to Resource

File: app/services/code_climate/update_project_service.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    def call
+      return unless update_metric? && code_climate_project_info
+
+      create_project_code_climate_metric unless project_code_climate_metric
+
+      project_code_climate_metric.update!(
+        code_climate_rate: code_climate_project_info.rate,
+        invalid_issues_count: code_climate_project_info.invalid_issues_count,
+        wont_fix_issues_count: code_climate_project_info.wont_fix_issues_count
```
Comment (santiagovidal) [inline app/services/code_climate/update_project_service.rb:17]: Can we also include the number of open issues? Sorry, I forgot to add it to the wireframe 😅

File: spec/services/code_climate/update_all_projects/code_climate_update_all_projects_service_spec.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  subject { CodeClimate::UpdateAllProjectsService }
+  let(:update_all_projects_code_climate_info) { subject.call }
+
+  it 'with no projects does no fail' do
+    expect { update_all_projects_code_climate_info }.not_to raise_error
+  end
+
+  context 'with a project' do
+    before do
+      first_project
```
Comment (santiagovidal) [inline spec/services/code_climate/update_all_projects/code_climate_update_all_projects_service_spec.rb:13]: You can use `let!` below and remove this block

File: spec/services/code_climate/update_all_projects/code_climate_update_all_projects_service_spec.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+    it('calls UpdateProjectService on that project') do
+      expect(CodeClimate::UpdateProjectService).to receive(:call).once.with(first_project)
+      update_all_projects_code_climate_info
+    end
+  end
+
+  context 'with many projects' do
+    before do
+      first_project
```
Comment (santiagovidal) [inline spec/services/code_climate/update_all_projects/code_climate_update_all_projects_service_spec.rb:26]: Same here.

File: app/services/code_climate/update_project_service.rb (hunk 5 above)
Comment (mrubi-rootstrap) [reply app/services/code_climate/update_project_service.rb:17]: Yes, it should be possible.
I have a question though.
CodeClimate issue status are `[new, open, confirmed, invalid, wontfix]`. I don't know what's the difference between `new` and `open`. Would `#new + #count + #confirmed`be  the number you are looking for?

**Comment by mrubi-rootstrap** [comment]
> Very nice! We can even extract your code and create a gem 😄

On that event a better error handler of the CodeClimate API response should be used. For this app returning nil from != 200 is fine but I think a gem should have some sort of error handler or response block to let the caller to  query the response code and evaluate a custom action

File: app/services/code_climate/api/client.rb [hunk 8]
```diff
@@ -0,0 +1,96 @@
+module CodeClimate
```
Comment (horacio) [inline app/services/code_climate/api/client.rb:1]: Nice! Reminds me to the Maybe monad.

File: app/services/code_climate/api/client.rb [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+      def snapshot_remote_query(repo_id:, snapshot_id:)
+        RemoteQuery.new("repos/#{repo_id}/snapshots/#{snapshot_id}")
+      end
+
+      def snapshot_issues_remote_query(repo_id:, snapshot_id:)
+        RemoteQuery.new("repos/#{repo_id}/snapshots/#{snapshot_id}/issues")
+      end
+
+      def safely
```
Comment (horacio) [inline app/services/code_climate/api/client.rb:64]: s/safely/silently/g 😆

**Comment by horacio** [comment]
Awesome job, @mrubi-rootstrap!

File: spec/services/code_climate/update_all_projects/code_climate_update_all_projects_service_spec.rb (hunk 6 above)
Comment (mrubi-rootstrap) [reply spec/services/code_climate/update_all_projects/code_climate_update_all_projects_service_spec.rb]: Fixed it. Thanks!

File: spec/services/code_climate/update_all_projects/code_climate_update_all_projects_service_spec.rb (hunk 7 above)
Comment (mrubi-rootstrap) [reply spec/services/code_climate/update_all_projects/code_climate_update_all_projects_service_spec.rb]: Fixed it. Thanks!

File: app/services/code_climate/update_project_service.rb (hunk 5 above)
Comment (mrubi-rootstrap) [reply app/services/code_climate/update_project_service.rb:18]: Done the backend task.
I've added a card to the backlog with the task for showing it in the frontend.

File: app/services/code_climate/api/client.rb (hunk 8 above)
Comment (mrubi-rootstrap) [reply app/services/code_climate/api/client.rb:1]: Had to google it :)

File: app/services/code_climate/api/client.rb (hunk 9 above)
Comment (mrubi-rootstrap) [reply app/services/code_climate/api/client.rb:64]: Yes, you are right.
I will change it to `ignoring_raised_errors` to make its intention more clear

File: spec/services/code_climate/code_climate_spec_helper.rb [hunk 10]
```diff
@@ -0,0 +1,24 @@
+require 'rails_helper'
+
+module CodeClimateSpecHelper
```
Comment (hvilloria) [inline spec/services/code_climate/code_climate_spec_helper.rb:3]: if this is a mock, it should be inside `spec/support/helpers/` and be called something like `code_climate_api_mocker`

Comment (mrubi-rootstrap) [reply spec/services/code_climate/code_climate_spec_helper.rb]: Done it. Thanks!

**Pull request merged**